src/Components/Server/src/ComponentHub.cs

```diff
@@ -66,7 +66,31 @@ public override Task OnDisconnectedAsync(Exception exception)
             }
 
             CircuitHost = null;
-            return _circuitRegistry.DisconnectAsync(circuitHost, Context.ConnectionId);
+            if (exception != null)
+            {
+                return _circuitRegistry.DisconnectAsync(circuitHost, Context.ConnectionId);
+            }
+            else
+            {
+                // The client will gracefully disconnect when using websockets by correctly closing the TCP connection.
+                // This happens when the user closes a tab, navigates away from the page or reloads the page.
+                // In these situations we know the user is done with the circuit, so we can get rid of it at that point.
+                // This is important to be able to more efficiently manage resources, specially memory.
+                return TerminateCircuitGracefully(circuitHost);
+            }
+        }
+
+        private async Task TerminateCircuitGracefully(CircuitHost circuitHost)
+        {
+            try
+            {
+                Log.CircuitTerminatedGracefully(_logger, circuitHost.CircuitId);
+                await circuitHost.DisposeAsync().AsTask();
```

Comment:
Out of interest, do we need the `.AsTask()` here?

Reply:
Do we need to remove the circuit from the active circuit registry?

Reply:
> Out of interest, do we need the `.AsTask()` here?

Its an artifact of the refactor.

> Do we need to remove the circuit from the active circuit registry?

Probably, I missed that part.